snapcraft/plugins/_ros/ros2.py

```diff
@@ -0,0 +1,181 @@
+    repo,
+    sources,
+)
+
+logger = logging.getLogger(__name__)
+
+_ROS2_URL_TEMPLATE = (
+    'https://raw.githubusercontent.com/ros2/ros2/{version}/ros2.repos'
+)
+
+
+_INSTALL_TOOLS_STEP = 'install-tools'
+_FETCH_ROS2_STEP = 'fetch-ros2'
+_BUILD_ROS2_STEP = 'build-ros2'
+
+
+class Bootstrapper:
+    """Bootstrap ROS2 by building the underlay from source."""
+
+    def __init__(self, *, version, bootstrap_path, ubuntu_sources, project):
+        """Initialize bootstrapper.
+
+        :param str version: The ROS2 version to boostrap (e.g. release-beta3)
+        :param str bootstrap_path: Working directory for the bootstrap process
+        :param str ubuntu_sources: Ubuntu repositories from which dependencies
+                                   will be fetched.
+        :param project: Instance of ProjectOptions for project-wide settings
+        :type project: snapcraft.ProjectOptions
+        """
+        self._version = version
+        self._ubuntu_sources = ubuntu_sources
+        self._project = project
+
+        self._bootstrap_path = bootstrap_path
+        self._tool_dir = os.path.join(self._bootstrap_path, 'tools')
+        self._state_dir = os.path.join(self._bootstrap_path, 'state')
+        self._underlay_dir = os.path.join(self._bootstrap_path, 'underlay')
+        self._install_dir = os.path.join(self._bootstrap_path, 'install')
+        self._build_dir = os.path.join(self._bootstrap_path, 'build')
+        self._source_dir = os.path.join(self._underlay_dir, 'src')
+
+    def get_build_packages(self):
+        """Return the packages required for building the underlay."""
+
+        packages = ['cmake']
+
+        # Dependencies for FastRTPS
+        packages.extend(['libasio-dev', 'libtinyxml2-dev'])
+
+        # Dependencies for the rest of ros2
+        packages.extend([
+            'libopencv-dev', 'libpoco-dev', 'libpocofoundation9v5',
+            'libpocofoundation9v5-dbg', 'python3-dev', 'python3-empy',
+            'python3-nose', 'python3-pip', 'python3-setuptools',
+            'python3-yaml', 'libtinyxml-dev', 'libeigen3-dev'])
+
+        return packages
+
+    def get_stage_packages(self):
+        """Return the packages required for running the underlay."""
+
+        # Ament is a python3 tool, and it requires setuptools at runtime
+        return ['python3', 'python3-setuptools']
+
+    def pull(self):
+        """Download the ROS2 underlay.
+
+        This first downloads the tools required to fetch the underlay, and then
+        uses those tools fetch the underlay. Both steps are run in a manner
+        such that they are only run once, even if subsequent steps fail.
+        """
+        self._run_step(
+            self._install_tools,
+            step=_INSTALL_TOOLS_STEP,
+            skip_message='Tools already installed. Skipping...')
+        self._run_step(
+            self._fetch_ros2,
+            step=_FETCH_ROS2_STEP,
+            skip_message='ros2 already fetched. Skipping...')
```

Comment:
Maybe, the `Skipping...` part could be added by _run_step. That way it will be harder to fail to follow the convention.

Reply:
I'd rather keep it simple, otherwise I need to start asking "should it add the period at the end as well?" And then it's not at all clear how the function is supposed to be used, either.